File: Medium/3612_Process_String_with_Special_Operations_I/Solution.cpp
#include "../../devlibs/cpp/cpphelper.h"

using namespace std;


// NOTE:
// For LeetCode submission, copy only the `class Solution` part.


class Solution {
public:
    string processStr(string s) {
        string result;

        for (char c : s) {
            if (c == '*') {
                // Remove the last character
                if (!result.empty()) {
                    result.pop_back();
                }
            } else if (c == '#') {
                // Duplicate
                if (!result.empty()) {
                    result += result;
                }
            } else if (c == '%') {
                // Reverse
                reverse(result.begin(), result.end());
            } else {
                // Append
                result += c;
            }
        }

        return result;
    }
};


int main(int argc, char* argv[]) {
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    Solution solution;

    // test cases 1
    string s1 = "a#b%*";
    cout << solution.processStr(s1) << endl;

    // test cases 2
    string s2 = "z*#";
    cout << solution.processStr(s2) << endl;

    return 0;
}

File: Medium/3612_Process_String_with_Special_Operations_I/Solution.java



public class Solution {
    public String processStr(String s) {
        StringBuilder result = new StringBuilder();

        for (char c : s.toCharArray()) {
            if (c == '*') {
                // Remove the last character
                if (result.length() > 0) {
                    result.deleteCharAt(result.length() - 1);
                }
            } else if (c == '#') {
                // Duplicate
                if (result.length() > 0) {
                    result.append(result.toString());
                }
            } else if (c == '%') {
                // Reverse
                result.reverse();
            } else {
                // Append
                result.append(c);
            }
        }

        return result.toString();
    }

    public static void main(String[] args) {
        Solution solution = new Solution();

        // test cases 1
        String s1 = "a#b%*";
        System.out.println(solution.processStr(s1));

        // test cases 2
        String s2 = "z*#";
        System.out.println(solution.processStr(s2));
        
    }
}